src/report/chunked.rs

```diff
@@ -0,0 +1,992 @@
+//! A wire protocol for representing ekotrace log reports
+//! that are fragmented into multiple chunks due to sizing
+//! constraints
+//!
+use crate::compact_log::CompactLogItem;
+use crate::history::DynamicHistory;
+use crate::ReportError;
+use crate::TracerId;
+use core::borrow::Borrow;
+use core::mem::{align_of, size_of, MaybeUninit};
+use static_assertions::{assert_eq_align, assert_eq_size, const_assert_eq, const_assert_ne};
+
+/// struct that represents a single report chunk
+/// as it would appear on the wire.
+///
+/// Report chunks have a maximum size of 256 bytes
+/// Report chunks ought to be aligned to 4 since their
+/// payload (in the most common log data type case) is
+/// often best interpreted as 4-byte u32s. However,
+/// this alignment is not mandatory.
+///
+/// Numeric fields are encoded in little endian format.
+#[repr(C, align(1))]
+pub struct WireChunk {
+    /// Fixed-sized always-fully-initialized header portion of the WireChunk format
+    pub header: WireChunkHeader,
+    /// The content of the report chunk, only certain to be initialized
+    /// after a write_next_report_chunk up to offset `header.n_chunk_payload_bytes`.
+    pub payload: [MaybeUninit<u8>; MAX_PAYLOAD_BYTES_PER_CHUNK],
+}
+assert_eq_size!([u32; MAX_CHUNK_U32_WORDS], WireChunk);
+
+/// Fixed-sized always-initialized header portion of the WireChunk format
+#[repr(C, align(1))]
+pub struct WireChunkHeader {
+    /// A magical (constant) value used as a hint about the
+    /// data encoded in this pile of bytes.
+    pub fingerprint: [u8; 4],
+    /// A u32 representing the tracer_id (a.k.a. location id) of the
+    /// ekotrace agent instance producing this report.
+    pub location_id: [u8; 4],
+    /// A u16 representing the group of report chunks to which
+    /// this chunk belongs. Determined by the ekotrace instance.
+    /// Expected, but not guaranteed to be increasing
+    /// and wrapping-overflowing during the lifetime of an ekotrace
+    /// instance.
+    ///
+    /// Erring on the side of a large
+    /// number of these to support cases like:
+    ///   * a long-lasting partition from the report collector whilst reports
+    /// are archived for future sending
+    ///   * a very high report production cadence relative to the transmission cadence
+    pub chunk_group_id: [u8; 2],
+    /// In what ordinal position does this chunk's payload live relative to the other chunks.
+    pub chunk_index: [u8; 2],
+    /// Does this chunk contain log data (0001) or extension data (0010)?
+    pub chunk_data_type: u8,
```

Comment:
This could also be represented, somewhat more compactly, with a third fingerprint value up-top. Any particular reason to do it in the body here?

Reply:
Because the main point of the framing fingerprint is to sanity-check that the following data should be interpreted in a particular way. The interpretation of the header does not change based on the content of the payload.  I think a better name for this field might be "payload_data_type" rather than "chunk_data_type", on further thought, so I understand the conflation.  Will rename.

An additional reason is that we'd like to be able to use the same framing format for log data and extension data, so if the framing is the same, then I think it makes sense to keep the same fingerprint.

Finally, hyper-compaction is not yet a goal, so the nicety of keeping things well aligned won out. Presently the payload at a nice offset (16 bytes in, so also 4 aligned), which is a nice perk for potential optimization later.